streamz/tests/test_sources.py

```diff
@@ -95,3 +95,10 @@ def test_process():
     s.start()
     yield await_for(lambda: out == [b'0\n', b'1\n', b'2\n', b'3\n'], timeout=5)
     s.stop()
+
+
+def test_from_iterable():
+    source = Source.from_iterable(range(3))
+    L = source.sink_to_list()
+    source.start()
+    wait_for(lambda: L == [0, 1, 2], 0.1)
```

Comment:
Is there a lag, or is this immediate?

Reply:
There's a lag. I tried doing just `assert L == [0, 1, 2]` and it fails: I think the callback is not run immediately at `.start()` and the interpreter runs the assertion first.

Reply:
Right - it was not immediately obvious to me that there *should* be a lag, which is a consequence of emitting asynchronously. Would you say that this basically comes from wanting to respect backpressure? 

I would imagine, if this class were not async, that backpressure would happen in another node downstream that *is*, such as `buffer`, which would take responsibility for creating the loop, if it doesn't exist. Then, calls to `emit` on the source would block until all items are processed.

Reply:
> Would you say that this basically comes from wanting to respect backpressure?

Yes, but also see below.

> I would imagine, if this class were not async, that backpressure would happen in another node downstream that is

There's a couple of things.

- `Source` interface should be uniform. You have to call `.start()`, the call doesn't block and exits immediately. All sources are running in a concurrent fashion.
- The source is what's "running" the pipeline, so there has to be some code that runs some steps, emits and yields control so that other nodes can do their thing too. This is what async coroutines and plain generators can do, but in case of generators, something has to run the generator. That's what the event loop is for, so we come back to `gen.coroutine`s.
- I'm not sure how we could achieve this without it being async. We could run `loop.run_in_executor` in `.start()`, or run it without the loop in a thread, but it's really the same thing, just uglier (and breaks backpressure, I checked). If `.start()` blocks, this breaks the uniformity and is even more unexpected.

So I don't know what it _means_ for a source to not be async. If you have some specific solution in mind, please point me in the right direction.

Reply:
Added a test for backpressure, in case you (or someone) wants to experiment with this not being async.